### PR TITLE
Tests for cluster features

### PR DESCRIPTION
**Pull request by piodul**
This PR adds a set of tests for cluster features. The set covers those tests from the test plan for the upcoming "cluster features on raft" that do not depend on implementation-specific details. Those tests are applicable to the existing gossip-based implementation, so they can be useful for us right now.

The tests simulate cluster upgrades by conditionally marking support for the TEST_ONLY_FEATURE on the nodes via error injection. Therefore, the tests work only in non-release mode.

The `test_partial_upgrade_can_be_finished_with_removenode` test is marked as skipped because of a bug in gossip that prevents features from being enabled if a node was removed within the last 3 days (#14194).

### CODE REVIEW

File: test/pylib/util.py [hunk 1]
```diff
@@ -121,4 +121,28 @@ async def read_barrier(cql: Session, host: Host):
...
 
+# Wait for the given feature to be enabled.
+async def wait_for_feature(feature: str, cql: Session, host: Host, deadline: float) -> None:
+    async def feature_is_enabled():
+        rs = await cql.run_async("select value from system.scylla_local where key = 'enabled_features'", host=host)
+        if rs:
+            value = rs[0].value
+            if feature in value:
+                return True
+        return None
```
Comment (michoecho) [inline test/pylib/util.py:132]: Note: could be shortened to:
```
return feature in get_enabled_features(cql, host) or None
```

File: test/topology/test_cluster_features.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+    # The feature should become enabled on all nodes soon.
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+    logging.info(f"Waiting until {TEST_FEATURE_NAME} is enabled on all nodes")
+    await asyncio.gather(*(wait_for_feature(TEST_FEATURE_NAME, cql, h, time.time() + 60) for h in hosts))
+
+
+@pytest.mark.asyncio
+async def test_downgrade_after_partial_upgrade(manager: ManagerClient) -> None:
+    """Simulates an partial of a cluster by enabling the test features in all
```
Comment (michoecho) [inline test/topology/test_cluster_features.py:96]: "an partial" -> "a partial upgrade"
?

**Review by michoecho — APPROVED** [review]

File: test/pylib/util.py (hunk 1 above)
Comment (piodul) [reply test/pylib/util.py]: Good point, done

File: test/topology/test_cluster_features.py (hunk 2 above)
Comment (piodul) [reply test/topology/test_cluster_features.py]: Fixed, thanks

**Comment by piodul** [comment]
v2:
- Moved `get_supported_features`/`get_enabled_features` to a separate commit
- Reimplemented some parts of `wait_for_feature` via `get_enabled_features`
- Fixed a typo in a comment

**Review by kbr-scylla — APPROVED** [review]

**Comment by piodul** [comment]
> CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1774/

It looks like sanity tests failed due to a spot termination. I'll rerun.

File: test/pylib/util.py [hunk 3]
```diff
@@ -124,13 +124,21 @@ async def read_barrier(cql: Session, host: Host):
...
+async def get_supported_features(cql: Session, host: Host) -> set[str]:
+    """Returns a set of cluster features that a node advertises support for."""
+    rs = await cql.run_async(f"SELECT supported_features FROM system.local WHERE key = 'local'", host=host)
+    return set(rs[0].supported_features.split(","))
+
+
+async def get_enabled_features(cql: Session, host: Host) -> set[str]:
+    """Returns a set of cluster features that a node considers to be enabled."""
+    rs = await cql.run_async(f"SELECT value FROM system.scylla_local WHERE key = 'enabled_features'", host=host)
+    return set(rs[0].value.split(","))
```
Comment (nyh) [inline test/pylib/util.py:141]: The code before this patch didn't seem to have the split()... Was it simply wrong?

**Review by nyh — APPROVED** [review]

File: test/pylib/util.py (hunk 3 above)
Comment (michoecho) [reply test/pylib/util.py:141]: If I understand you correctly, you mean the fact that before it was `string1 in string2` and now it's `string1 in string2.split(",")`?

The code before the patch was wrong, but it was OK as long as no feature name was a substring of another feature name. (Note that in Python `string1 in string2` checks whether `string1` is a substring of `string1`. So before the patch, we would check whether the feature name is a substring of the comma-separated string of features. After the patch, we explicitly construct the list (Python list) of features from the comma-separated string, and check whether the feature is in the list.

Comment (nyh) [reply test/pylib/util.py:141]: Right, I forgot that "string in string" also "sort of, sometimes" works. Of course the new version with "string in set" is better.

Comment (piodul) [reply test/pylib/util.py:141]: It used to work, but it was less explicit. `str1 in str2` in python is equivalent to "str1 is a substring of str2", so previously the function `wait_for_feature` checked whether the feature name is a _substring_ of the comma-separated string of features.

Comment (piodul) [reply test/pylib/util.py:141]: > The code before the patch was wrong, but it was OK as long as no feature name was a substring of another feature name.

Right, I missed that possibility. So the previous version wasn't entirely correct after all.

**Pull request merged**